src/awkward/operations/ak_local_index.py

```diff
@@ -88,8 +93,35 @@ def local_index(array, axis=-1, *, highlevel=True, behavior=None, attrs=None):
 
 
 def _impl(array, axis, highlevel, behavior, attrs):
-    axis = regularize_axis(axis)
     with HighLevelContext(behavior=behavior, attrs=attrs) as ctx:
         layout = ctx.unwrap(array, allow_record=False, primitive_policy="error")
+
+    axis = regularize_axis(axis)
+
+    # Handle named axis
+    named_axis = _get_named_axis(ctx)
+    # Step 1: Normalize named axis to positional axis
+    axis = _named_axis_to_positional_axis(named_axis, axis)
+
+    if not is_integer(axis):
+        raise TypeError(f"'axis' must be an integer by now, not {axis!r}")
```

Comment:
This looks like more of an assertion. It's a good idea to only use exceptions like TypeError and ValueError for user (or dependent library) mistakes and AssertionError and RuntimeError for mistakes in Awkward itself.

Reply:
```suggestion
    assert is_integer(axis), f"'axis' must be an integer by now, not {axis!r}"
```

Reply:
These checks exist like this already in awkward-array/main: https://github.com/scikit-hep/awkward/blob/main/src/awkward/operations/ak_is_none.py#L48-L49
I just added them consistently to every high-level operation.

In principle this is the error a user sees when they pass in an axis that is neither a valid `named_axis` or an `int` (or sometimes `None`) - so it should be a user mistake.

I thought of it as 3 possibilities:
1. it is a valid `named_axis` (a `str`): translate it to `int` if possible else raise a meaningful error for the user
2. it is `int` (or sometimes `None`): just pass through
3. it is something else, e.g. a `dict` (why would someone ever do this, but it is not prohibited): raise the `TypeError` as above